### PR TITLE
treat compilation warnings as errors

### PR DESCRIPTION
**Pull request by vladak**
# Suppressions

Here's list of types of problems that had to be suppressed:

## `rawtypes`
e.g. `ConfigurationHelp` contains the `getSetters(Class klass)` definition where `Class` stands by itself (i.e. does not have any qualifier in the form of `Class<T>`) which will trigger
```
Warning:(108, 44) java: found raw type: java.lang.Class
  missing type arguments for generic class java.lang.Class<T>
```

## `unchecked`
e.g. `RuntimeEnvironment#getProjects()`:

```java
public Map<String,Project> getProjects() {
        return (Map<String, Project>)getConfigurationValue("projects");
    }
```
which will result in:
```
Warning:(476, 59) java: unchecked cast
  required: java.util.Map<java.lang.String,org.opengrok.indexer.configuration.Project>
  found:    java.lang.Object
```
Also described on https://stackoverflow.com/questions/262367/type-safety-unchecked-cast

I was hoping that the suggestion https://github.com/oracle/opengrok/pull/2423#discussion_r226255985 could help to fix that however the type cast is still there (besides it does not seem possible to write something like `Set<Group>.class`). According to https://stackoverflow.com/questions/509076/how-do-i-address-unchecked-cast-warnings it should be possible to suppress this only for local variables however I could not make it work and it is actually not strictly necessary given most of the involved methods are one liners.

### unchecked varargs
A sub-type of this warning happens because using varargs method such as `contains()` (from Hamcrest) in `SuggesterControllerTest`:
```java
        assertThat(res, contains(new SimpleEntry<>("topclass", 15), new SimpleEntry<>("mynamespace", 10)));

```
which results in this warning:
```
Warning:(611, 33) java: unchecked generic array creation for varargs parameter of type java.util.Map.Entry<java.lang.String,java.lang.Integer>[]
```
This is explained on https://stackoverflow.com/questions/22521288/what-are-auxiliary-classes and also in detail on http://www.angelikalanger.com/GenericsFAQ/FAQSections/ProgrammingIdioms.html#FAQ300 :

> This warning is confusing because there is no array creation expression anywhere in the source code. In order to understand, what the compiler complains about you need to keep in mind two things:
> -   Variable argument lists are translated by the compiler into an array.
> -   Creation of arrays with a non-reifiable component type is not permitted.

To paraphrase `SimpleEntry<String, int>` is 
> not a reifiable type, that is, it loses information as a result of type erasure and is at runtime represented as the raw type SimpleEntry instead of the exact type SimpleEntry<String,int> .  The loss of information leads to problems with arrays of such non-reifiable component types.

## `deprecation`
Existing code uses some APIs marked as deprecated in some of the 3rd party libraries, e.g. Lucene's `RAMDirectory` that cannot be fixed right away (see #2454).

## `fallthrough`
Some of the switch-case statements have fall-through that look deliberate.

## `auxiliaryclass`
Had to add this to the overall lint flags since the generated classes contain auxiliary classes (https://stackoverflow.com/questions/22521288/what-are-auxiliary-classes).

# Fixups

## missing qualifiers
Some of the methods were missing qualifiers, e.g. `DummyHttpServletRequestLdap` implements the `HttpServletRequest` interface and contained overrides like:
```java
public Enumeration getAttributeNames()
```
where generic type is missing its qualifier. In this case it should say `Enumeration<String>`. If missing, IDEA will actually generate 2 warnings:
```
Warning:(318, 24) java: getAttributeNames() in opengrok.auth.plugin.util.DummyHttpServletRequestLdap implements getAttributeNames() in javax.servlet.ServletRequest
  return type requires unchecked conversion from java.util.Enumeration to java.util.Enumeration<java.lang.String>
Warning:(318, 12) java: found raw type: java.util.Enumeration
  missing type arguments for generic class java.util.Enumeration<E>
```

A sub-case of this could be missing qualifier that could be inferred like in the case of calling `prepareIndexer()` with `new ArrayList(repositories)` as argument. Using empty diamond operator fixed this problem.

## superfluous casting
`FileHistoryCache#mergeOldAndNewHistory()` had this:
```java
                ListIterator li = listNew.listIterator(listNew.size());
                while (li.hasPrevious()) {
                    listOld.add(0, (HistoryEntry) li.previous());
                }
                history = new History(listOld);
```
which could be avoided by using qualifier for `ListIterator`.

## calling static methods using instance
`GitHistoryParser` had this:
```java
    public void processStream(InputStream input) throws IOException {
        try (BufferedReader in = new BufferedReader(repository.newLogReader(input))) {
```
where `repository` was instance of `GitRepository`. This produced this warning:
```
Warning:(76, 63) java: static method should be qualified by type name, org.opengrok.indexer.history.GitRepository, instead of by an expression
```

### avoiding Thread.currentThread()
`Executor` used `Thread.currentThread().getDefaultUncaughtExceptionHandler()`which access static method so `currentThread()` is not necessary otherwise it leads to:
```
Warning:(384, 35) java: static method should be qualified by type name, java.lang.Thread, instead of by an expression
```

## using getDeclaredConstructor()
`clazz.newInstance()` is not enough in newer JDKs, `clazz.getDeclaredConstructor().newInstance()` has to be used instead since the former was deprecated in Java 9 (https://stackoverflow.com/questions/53257073/java-9-replace-class-newinstance)

## avoid using constructor for Integer/Long
Using `new Integer` is not necessary (and actually wastes memory) and can be replaced `Integer.valueOf()` (https://stackoverflow.com/questions/2974561/new-integer-vs-valueof)

### CODE REVIEW

**Comment by vladak** [comment]
So far trying to make everything clean was interesting experience within IDEA. It keeps reporting more and more failures for each build. Even after build came out clean, next rebuild lead to new warnings. Rinse and repeat multiple times and the builds are clean however it might be just a coincidence.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/web/DummyHttpServletRequest.java [hunk 1]
```diff
@@ -383,7 +383,7 @@ public Enumeration getParameterNames() {
     }
 
     @Override
-    public Map getParameterMap() {
+    public Map<String,String[]> getParameterMap() {
```
[the remark this answers is not in the document]
Comment (tulinkry) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/web/DummyHttpServletRequest.java]: there should be a space after the comma, no?

[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/web/DummyHttpServletRequest.java]: I see, in general probably yes although the cstyle check does not enforce this. That said, for return types more condensed version might look better.

[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/web/DummyHttpServletRequest.java]: The thing is the http://checkstyle.sourceforge.net/config_whitespace.html#GenericWhitespace module only checks white space around `<` and `>`.

**Pull request merged**